### ISSUE TITLE
Subscribe repeatedly

### ISSUE THREAD

**Issue by JokerQyou** [andycampbell92/mqtt-socketio#2]
It's me again :)
This bridge actually works and save me a lot of time to implement one myself, because I'm not quite a Node.js user and have no background knowledge on it. 
I thought I find a logical loophole in the codes. 
[Here]( https://github.com/andycampbell92/mqtt-socketio/blob/96585b6bbcb69f7d112d68dc886d3f64d3798a1f/mqtt-io-bridge.js#L16-L22 ), ascoltatore is subscribing a topic whenever a new client connects to this bridge and subscribe this topic. Yet I think it's very common for multiple clients to subscribe to one topic. And this would cause a problem: 

Say client A and client B are connecting through this bridge, and they both subscribe to topic 'test', then this bridge would subscribe 'test' twice. And if a message 'hi' was published to 'test', both A and B will receive the same message twice. This situation would be much more worse if there's more clients. 

I think by maintaining a `topic_pool` and `socket_pool` which stores the mapping of `<socket> => [<topic>, <topic>]` could prevent this problem. But it could probably consume more memory.

**Comment by andycampbell92** [comment]
Hmm just taken a look now sorry for not getting back sooner.

I think it may be better if we created a ascoltatore client for each socket.io connection because then we don't have to worry about maintaining a list. Although this would take quite a bit of memory I can imagine this kind of problem reoccurring.

The problem with maintaining our own subscription list is we risk recreating the functionality of our mqtt client ascoltatore.

**Comment by JokerQyou** [comment]
I agree with most part of your opinion, but I don't quite understand this: 

> Although this would take quite a bit of memory I can imagine this kind of problem reoccurring.

As far as I can see, maintaining our own lists would prevent this (subscribe repeatedly) from happening. 

I think the most shortage of using only one ascoltatore for the whole bridge is that all the clients' last will message would be lost if they disconnect unexpectedly. Of course we could make some promises between the clients and the bridge, say 'Send a last will message to the bridge right after a client connected, and the bridge will send it out to the broker if the client disconnect', but that would increase the complexity. 

And by using standalone ascoltatore for each client connection would increase the memory usage (I've not tested yet but I think it would consume more memory than the 'maintain-list' plan). 

I'll re-think about this if the 'maintain-list' plan come into problems. (Right now it works well and I'm busy with something else. :) )

**Comment by JokerQyou** [comment]
Alright... this did run into some creepy problems...
Even if the bridge managed to subscribe only once to each topic, and unsubscribe from every topic that no clients subscribed, it will still receive multiple duplicated messages. 

The detailed usage is as followed. 

An HTML page as the client, using the bridge to connect to a MQTT broker, the page subscribe a topic once it finished loading. And after refresh the page N times, the page will receive N messages whenever the broker send one message. 

I don't yet know whether it's because refreshing does not emit the `disconnect` event in Socket.io, or it's because something else inside ascoltatori. Just take a note here, in case someone else encountered this problem too. I'll take a look back if I have more time.

**Comment by andycampbell92** [comment]
Sorry I am in the process of moving house so I haven't had time to look at
the library yet.

The bug sounds a lot more obscure and might take a fair bit of
investigation.

**Comment by JokerQyou** [comment]
Hi @andycampbell92 , I've finally solved this problem some days ago (sorry I forgot to close this issue that time, I was a little busy). It turns out to be a `QoS` mismatching thing. I was using mosquitto as my MQTT broker when I first opened this issue, and was using QoS 2, but it seems ascoltatore only support QoS 0 and QoS 1, so the broker is sending the same messages again and again. 
How did I solve this? Well I changed the MQTT broker to Mosca for its flexibility to write custom Javascript to control the client. And Mosca does not support QoS 2 either (well, deep inside it uses ascoltatore), so I eventually have the broker and the client **both** set to QoS 1. And boom! Problem solved. 
Hope this issue being helpful for others who run into the same problem.
